spackbot/pr_add_reviewers.py

```diff
@@ -0,0 +1,235 @@
+"""
+
+
+async def changed_packages(gh, pull_request):
+    """Return an array of packages that were modified by a PR.
+
+    Ignore deleted packages, since we can no longer query them for
+    maintainers.
+
+    """
+    # see which files were modified
+    packages = []
+    async for f in gh.getiter(pull_request["url"] + "/files"):
+        filename = f["filename"]
+        status = f["status"]
+
+        if status == "removed":
+            #            print(f["patch"])
+            continue
+
+        match = re.match(package_path, filename)
+        if not match:
+            continue
+        packages.append(match.group(1))
+
+    return packages
+
+
+@contextlib.contextmanager
+def temp_dir():
+    """Create a temporary directory, cd into it, destroy it and cd back when done."""
+    pwd = os.getcwd()
+    with tempfile.TemporaryDirectory() as temp_dir:
+        try:
+            os.chdir(temp_dir)
+            yield temp_dir
+        finally:
+            os.chdir(pwd)
+
+
+def find_maintainers(packages, repository, pull_request, number):
+    """Return an arrays of packages with maintainers, an array of packages
+    without maintainers, and a set of maintainers.
+
+    Ignore the author of the PR, as they don't need to review their own PR.
+    """
+    author = pull_request["user"]["login"]
+    clone_url = repository["clone_url"]
+
+    with_maintainers = []
+    without_maintainers = []
+    maintainers = set()
+
+    with temp_dir() as cwd:
+        # Clone appropriate PR branch
+        git("clone", f"{clone_url}")
+        git("fetch", "origin", f"pull/{number}/head:PR{number}", _cwd="spack")
+        git("checkout", f"PR{number}", _cwd="spack")
+
+        # Add `spack` to PATH
+        os.environ["PATH"] = f"{cwd}/spack/bin:" + os.environ["PATH"]
+        from sh import spack
+
+        for package in packages:
+            logger.debug(f"Package: {package}")
+
+            # Query maintainers
+            pkg_maintainers = spack("maintainers", package, _ok_code=(0, 1)).split()
+            pkg_maintainers = set(pkg_maintainers)
+
+            logger.debug("Maintainers: %s" % ", ".join(sorted(pkg_maintainers)))
+
+            if not pkg_maintainers:
+                without_maintainers.append(package)
+                continue
+
+            # No need to ask the author to review their own PR
+            if author in pkg_maintainers:
+                pkg_maintainers.remove(author)
+
+            if pkg_maintainers:
+                with_maintainers.append(package)
+                maintainers |= pkg_maintainers
+
+    return with_maintainers, without_maintainers, maintainers
+
+
+async def found(coroutine):
+    """Wrapper for coroutines that returns None on 404, result otherwise."""
+    try:
+        return await coroutine
+    except gidgethub.HTTPException as e:
+        if e.status_code == 404:
+            return None
+        raise
+
+
+async def add_reviewers(gh, repository, pull_request, number):
```

Comment:
All this async and await stuff is really fancy. Someday I would like to learn how to write code like this.

Reply:
It wasn't that hard to get used to -- it's just lightweight threads.